File: citeindex/agents/memory.py
"""File-based JSONL memory persistence for chat threads.

Stores query + response pairs as JSONL.  Each entry is hashed into a
Merkle DAG per Summary.md.  PostgreSQL migration deferred to Tier 3.
"""

import json
import logging
import os
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

from .models import MemoryEntry, SCHEMA_VERSION

logger = logging.getLogger(__name__)


class MemoryStore:
    """JSONL-backed memory store with per-thread files."""

    def __init__(self, memory_dir: str = "corpus/.memory") -> None:
        self.memory_dir = os.path.abspath(memory_dir)
        os.makedirs(self.memory_dir, exist_ok=True)

    def save(
        self,
        thread_id: str,
        query: str,
        response: str,
        evidence_node_ids: Optional[List[str]] = None,
    ) -> MemoryEntry:
        from ..ingestion.deterministic import hash_payload

        timestamp = datetime.now(timezone.utc).replace(microsecond=0).isoformat()
        payload = f"{timestamp}|{query}|{response}"
        entry_hash = hash_payload(payload)

        entry = MemoryEntry(
            entry_id=entry_hash[:16],
            timestamp=timestamp,
            thread_id=thread_id,
            query=query,
            response=response,
            evidence_node_ids=evidence_node_ids or [],
            sha256=entry_hash,
        )

        path = self._thread_path(thread_id)
        with open(path, "a", encoding="utf-8") as f:
            f.write(json.dumps(entry.to_dict(), ensure_ascii=False, sort_keys=True))
            f.write("\n")

        logger.info("Memory saved: entry_id=%s thread=%s", entry.entry_id, thread_id)
        return entry

    def load_thread(self, thread_id: str) -> List[MemoryEntry]:
        path = self._thread_path(thread_id)
        if not os.path.exists(path):
            return []

        entries: List[MemoryEntry] = []
        with open(path, "r", encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                try:
                    data = json.loads(line)
                    entries.append(MemoryEntry(**data))
                except Exception:
                    logger.warning("Skipping malformed memory entry", exc_info=True)
        return entries

    def search(self, query: str, thread_id: Optional[str] = None) -> List[MemoryEntry]:
        """Simple keyword search across memory entries."""
        from .indexing import tokenize

        query_tokens = set(tokenize(query))
        if not query_tokens:
            return []

        results: List[tuple] = []

        if thread_id:
            threads = [thread_id]
        else:
            threads = self._list_threads()

        for tid in threads:
            entries = self.load_thread(tid)
            for entry in entries:
                entry_tokens = set(tokenize(entry.query + " " + entry.response))
                overlap = len(query_tokens & entry_tokens)
                if overlap > 0:
                    results.append((overlap, entry))

        results.sort(key=lambda x: -x[0])
        return [entry for _, entry in results]

    def build_merkle_dag(self, thread_id: str) -> Dict[str, Any]:
        """Build a Merkle DAG for a thread's memory entries."""
        from ..ingestion.deterministic import build_merkle_tree

        entries = self.load_thread(thread_id)
        leaf_hashes = [e.sha256 for e in entries if e.sha256]
        if not leaf_hashes:
            return {"root": "", "leaf_count": 0}
        return build_merkle_tree(leaf_hashes)

    def _thread_path(self, thread_id: str) -> str:
        safe_id = "".join(c if c.isalnum() or c in "-_" else "_" for c in thread_id)
        return os.path.join(self.memory_dir, f"{safe_id}.jsonl")

    def _list_threads(self) -> List[str]:
        threads: List[str] = []
        if not os.path.isdir(self.memory_dir):
            return threads
        for fname in sorted(os.listdir(self.memory_dir)):
            if fname.endswith(".jsonl"):
                threads.append(fname[:-6])
        return threads

File: citeindex-rs/crates/core/src/memory.rs
//! Rust-native memory manager — Merkle DAG for chat memory with
//! file-based persistence and optional PostgreSQL backend.
//!
//! Matches Phase 3.4 and `rust_core_orchestration.yaml → manage_memory`.

use crate::merkle::{build_merkle_tree, sha256_hex, MerkleTree};
use chrono::Utc;
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};
use std::fs::{self, OpenOptions};
use std::io::{BufRead, BufReader, Write};
use std::path::{Path, PathBuf};

/// A single memory entry persisted to JSONL.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MemoryEntry {
    pub entry_id: String,
    pub timestamp: String,
    pub thread_id: String,
    pub query: String,
    pub response: String,
    pub evidence_node_ids: Vec<String>,
    pub sha256: String,
}

impl MemoryEntry {
    pub fn new(
        thread_id: &str,
        query: &str,
        response: &str,
        evidence_node_ids: Vec<String>,
    ) -> Self {
        let timestamp = Utc::now().format("%Y-%m-%dT%H:%M:%S+00:00").to_string();
        let payload = format!("{}|{}|{}", timestamp, query, response);
        let hash = sha256_hex(&payload);
        let entry_id = hash[..16].to_string();

        Self {
            entry_id,
            timestamp,
            thread_id: thread_id.into(),
            query: query.into(),
            response: response.into(),
            evidence_node_ids,
            sha256: hash,
        }
    }
}

/// File-based JSONL memory store with Merkle DAG support.
pub struct MemoryStore {
    memory_dir: PathBuf,
    legacy_memory_dir: Option<PathBuf>,
}

impl MemoryStore {
    pub fn new(memory_dir: &Path, legacy_memory_dir: Option<&Path>) -> Self {
        fs::create_dir_all(memory_dir).ok();
        Self {
            memory_dir: memory_dir.to_path_buf(),
            legacy_memory_dir: legacy_memory_dir.map(Path::to_path_buf),
        }
    }

    /// Save a memory entry to the thread's JSONL file.
    pub fn save(&self, entry: &MemoryEntry) -> anyhow::Result<()> {
        let path = self.thread_path(&entry.thread_id);
        let mut file = OpenOptions::new().create(true).append(true).open(&path)?;
        let line = serde_json::to_string(entry)?;
        writeln!(file, "{}", line)?;
        tracing::info!(
            entry_id = %entry.entry_id,
            thread = %entry.thread_id,
            "Memory entry saved"
        );
        Ok(())
    }

    /// Load all entries for a thread.
    pub fn load_thread(&self, thread_id: &str) -> Vec<MemoryEntry> {
        let mut by_id = HashMap::new();

        if let Some(legacy_dir) = self.legacy_memory_dir.as_ref() {
            for entry in self.load_thread_from_dir(legacy_dir, thread_id) {
                by_id.insert(entry.entry_id.clone(), entry);
            }
        }
        for entry in self.load_thread_from_dir(&self.memory_dir, thread_id) {
            by_id.insert(entry.entry_id.clone(), entry);
        }

        let mut entries = by_id.into_values().collect::<Vec<_>>();
        entries.sort_by(|a, b| a.timestamp.cmp(&b.timestamp));
        entries
    }

    /// Simple keyword search across all threads (or a specific thread).
    pub fn search(&self, query: &str, thread_id: Option<&str>) -> Vec<MemoryEntry> {
        let query_lower = query.to_lowercase();
        let keywords: Vec<&str> = query_lower.split_whitespace().collect();
        if keywords.is_empty() {
            return Vec::new();
        }

        let threads = match thread_id {
            Some(tid) => vec![tid.to_string()],
            None => self.list_threads(),
        };

        let mut results: Vec<(usize, MemoryEntry)> = Vec::new();
        for tid in &threads {
            for entry in self.load_thread(tid) {
                let text = format!("{} {}", entry.query, entry.response).to_lowercase();
                let hits = keywords.iter().filter(|kw| text.contains(*kw)).count();
                if hits > 0 {
                    results.push((hits, entry));
                }
            }
        }

        results.sort_by(|a, b| b.0.cmp(&a.0));
        results.into_iter().map(|(_, e)| e).collect()
    }

    /// Build a Merkle tree for all entries in a thread.
    pub fn build_merkle_dag(&self, thread_id: &str) -> MerkleTree {
        let entries = self.load_thread(thread_id);
        let hashes: Vec<String> = entries.iter().map(|e| e.sha256.clone()).collect();
        build_merkle_tree(&hashes)
    }

    /// List all thread IDs.
    pub fn list_threads(&self) -> Vec<String> {
        let mut threads = HashSet::new();
        if let Some(legacy_dir) = self.legacy_memory_dir.as_ref() {
            self.collect_thread_names(legacy_dir, &mut threads);
        }
        self.collect_thread_names(&self.memory_dir, &mut threads);

        let mut sorted = threads.into_iter().collect::<Vec<_>>();
        sorted.sort();
        sorted
    }

    fn load_thread_from_dir(&self, dir: &Path, thread_id: &str) -> Vec<MemoryEntry> {
        let path = self.thread_path_in_dir(dir, thread_id);
        if !path.exists() {
            return Vec::new();
        }

        let file = match fs::File::open(&path) {
            Ok(file) => file,
            Err(_) => return Vec::new(),
        };

        BufReader::new(file)
            .lines()
            .filter_map(|line| line.ok())
            .filter(|line| !line.trim().is_empty())
            .filter_map(|line| serde_json::from_str::<MemoryEntry>(&line).ok())
            .collect()
    }

    fn collect_thread_names(&self, dir: &Path, threads: &mut HashSet<String>) {
        if let Ok(entries) = fs::read_dir(dir) {
            for entry in entries.flatten() {
                if let Some(name) = entry.file_name().to_str() {
                    if name.ends_with(".jsonl") {
                        threads.insert(name.trim_end_matches(".jsonl").to_string());
                    }
                }
            }
        }
    }

    fn thread_path(&self, thread_id: &str) -> PathBuf {
        self.thread_path_in_dir(&self.memory_dir, thread_id)
    }

    fn thread_path_in_dir(&self, dir: &Path, thread_id: &str) -> PathBuf {
        let safe: String = thread_id
            .chars()
            .map(|c| {
                if c.is_alphanumeric() || c == '-' || c == '_' {
                    c
                } else {
                    '_'
                }
            })
            .collect();
        dir.join(format!("{}.jsonl", safe))
    }
}

#[cfg(test)]
mod tests {
    use std::env;

    use super::*;

    #[test]
    fn test_memory_store_merges_primary_and_legacy_entries_by_entry_id() {
        let root = env::temp_dir().join(format!("citeindex-memory-store-{}", uuid::Uuid::new_v4()));
        let primary = root.join("primary");
        let legacy = root.join("legacy");
        fs::create_dir_all(&primary).unwrap();
        fs::create_dir_all(&legacy).unwrap();

        let legacy_entry = MemoryEntry {
            entry_id: "mem-1".into(),
            timestamp: "2026-04-03T10:00:00+00:00".into(),
            thread_id: "thread-1".into(),
            query: "Old query".into(),
            response: "Old response".into(),
            evidence_node_ids: vec![],
            sha256: "old-hash".into(),
        };
        let primary_entry = MemoryEntry {
            entry_id: "mem-1".into(),
            timestamp: "2026-04-03T11:00:00+00:00".into(),
            thread_id: "thread-1".into(),
            query: "New query".into(),
            response: "New response".into(),
            evidence_node_ids: vec!["doc-1:node-1".into()],
            sha256: "new-hash".into(),
        };

        fs::write(
            legacy.join("thread-1.jsonl"),
            format!("{}\n", serde_json::to_string(&legacy_entry).unwrap()),
        )
        .unwrap();
        fs::write(
            primary.join("thread-1.jsonl"),
            format!("{}\n", serde_json::to_string(&primary_entry).unwrap()),
        )
        .unwrap();

        let store = MemoryStore::new(&primary, Some(&legacy));
        let entries = store.load_thread("thread-1");

        assert_eq!(entries.len(), 1);
        assert_eq!(entries[0].query, "New query");
        assert_eq!(entries[0].sha256, "new-hash");

        fs::remove_dir_all(root).unwrap();
    }
}
